File: Missing_Numbers.java

/*
    author:Karan
    created:26.03.2022 10:53:30
*/
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.StringTokenizer;
import java.util.*;

public class Missing_Numbers {
    static Integer A, B, C, D;

    static HashMap<Character, Integer> M = new HashMap<>();
    static HashMap<Character, Boolean> used = new HashMap<>();

    static Boolean calc(String s) {
        long add = M.get(s.charAt(0));
        long mul = M.get(s.charAt(1));
        long sub = M.get(s.charAt(2));
        long div = M.get(s.charAt(3));
        // System.out.println(sum + " " + mul + " " + sub + " " + div);
        long a = (add + sub) / 2;
        long b = (add - sub) / 2;
        // cout << a << " " << b << endl;
        if (a >= 1 && a <= 10000 && b >= 1 && b <= 10000 && mul == a * b && (div == a / b)) {
            x = a;
            y = b;
            return true;
        }
        return false;
    }

    static boolean ans;
    static long x, y;

    static String arr[] = new String[] { "ABCD", "ABDC", "ACBD", "ACDB", "ADBC", "ADCB", "BACD", "BADC", "BCAD", "BCDA",
            "BDAC", "BDCA", "CABD", "CADB", "CBAD", "CBDA", "CDAB", "CDBA", "DABC", "DACB", "DBAC", "DBCA", "DCAB",
            "DCBA" };

    static void solve() {
        for (String x : arr) {
            ans |= calc(x);
        }
    }

    public static void main(String[] args) throws IOException {
        try {
            FastScanner fs = new FastScanner();
            long T = fs.nextInt();
            for (long tt = 0; tt < T; tt++) {
                x = -1;
                y = -1;
                ans = false;
                A = fs.nextInt();
                B = fs.nextInt();
                C = fs.nextInt();
                D = fs.nextInt();
                M.clear();
                used.clear();
                for (char i = 'A'; i <= 'D'; i++) {
                    used.put(i, false);
                }
                M.put('A', A);
                M.put('B', B);
                M.put('C', C);
                M.put('D', D);
                solve();
                System.out.println(x + " " + y);
            }
        } catch (Exception e) {
            return;
        }
    }

    static class FastScanner {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer("");

        String next() {
            while (!st.hasMoreTokens())
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        int[] readArray(int n) {
            int[] a = new int[n];
            for (int i = 0; i < n; i++)
                a[i] = nextInt();
            return a;
        }

        long nextLong() {
            return Long.parseLong(next());
        }
    }
}

File: Missing_Numbers.cpp
/*
    author:Karan
    created:25.03.2022 23:34:37
*/
#if true
#pragma GCC target ("avx2")
#pragma GCC optimization ("O3")
#pragma GCC optimization ("unroll-loops")
#include<bits/stdc++.h>
using namespace std;
#define ll long long
#define all(ar) ar.begin(),ar.end()
#define endl '\n'
#define mset(a,x) memset(a,x,sizeof(a))
#endif
ll A, B, C, D;
map<char, ll> M;
string arr[] = { "ABCD", "ABDC", "ACBD", "ACDB", "ADBC", "ADCB", "BACD", "BADC", "BCAD", "BCDA",
            "BDAC", "BDCA", "CABD", "CADB", "CBAD", "CBDA", "CDAB", "CDBA", "DABC", "DACB", "DBAC", "DBCA", "DCAB",
            "DCBA" };
ll x, y;

bool calc(string s) {
    ll add = M[s[0]];
    ll mul = M[s[1]];
    ll sub = M[s[2]];
    ll div = M[s[3]];
    // cout << add << " " << mul << " " << sub << " " << div << endl;

    ll  a = (add + sub) / 2;
    ll  b = (add - sub) / 2;
    // cout << a << " " << b << endl;
    if (a >= 1 and a <= 10000 and b >= 1 and b <= 10000 and mul == a * b and (div == a / b)) {
        x = a;
        y = b;
        return true;
    }
    return false;

}
bool ans;
void k() {
    for (string x : arr) {
        ans |= calc(x);
    }
}
void solve() {
    cin >> A >> B >> C >> D;
    x = -1;
    y = -1;
    M.clear();
    M['A'] = A;
    M['B'] = B;
    M['C'] = C;
    M['D'] = D;
    k();
    cout << x << " " << y << endl;
}

int32_t main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    // freopen("input.txt","r",stdin);
    // freopen("output.txt","w",stdout);
    ll t = 1;
    cin >> t;
    while (t--) {
        solve();
    }
}